### ISSUE TITLE
HttpClient call to API crashes in Android Emulator

### ISSUE THREAD

**Issue by ElliottBrand** [dotnet/maui#1196]
### Description
Whenever I attempt to make an HttpClient call to an API, it crashes and closes the app in the Android emulator. I've heard there are some known issues with https, but I'm still getting it in http. 

I've added the service in a shared project

```
using System;
using System.Threading.Tasks;
using System.Net.Http;

namespace WebAssemblyApp.Shared
{
    public interface IAboutService
    {
        Task<string> GetAboutAsync();
    }

    public class AboutService : IAboutService
    {
        private readonly HttpClient _httpClient;

        public AboutService(HttpClient httpClient)
        {
            _httpClient = httpClient;
        }

        public async Task<string> GetAboutAsync()
        {
            try
            {
                var result = await _httpClient.GetAsync("api/about");
                if (result.IsSuccessStatusCode)
                {
                    return System.Text.Json.JsonSerializer.Serialize(result.Content.ReadAsStringAsync());
                }
                return "Error"; // Just notifying the status code wasn't success while debugging httpclient issue
            }
            catch(Exception ex)
            {
                return ex.Message;
            }
        }
    }
}
```

Added the ServiceCollectionExtensions to a Shared Components project, so the same Blazor/Razor components can be used in multiple projects

```
using System;
using Microsoft.Extensions.DependencyInjection;
using WebAssemblyApp.Shared;

namespace BlazorMaui
{
    public static class ServiceCollectionExtensions
    {
        public static IServiceCollection AddBlazorStuff(this IServiceCollection services, string baseUri)
        {
            services.AddHttpClient<IAboutService, AboutService>(httpClient => httpClient.BaseAddress = new Uri(baseUri));
            return services;
        }
    }
}
```
And set the baseUri as the non-secure url path to the API in the main MAUI startup, using the IP that Android maps to the localhost (navigating to http://10.0.2.2:5000/api/about in the Android web browser does show the content)

```
using Microsoft.AspNetCore.Components.WebView.Maui;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Maui;
using Microsoft.Maui.Hosting;
using Microsoft.Maui.Controls.Compatibility;

namespace BlazorMaui
{
	public class Startup : IStartup
	{
		public void Configure(IAppHostBuilder appBuilder)
		{
			appBuilder
				.UseFormsCompatibility()
				.RegisterBlazorMauiWebView(typeof(Startup).Assembly)
				.UseMicrosoftExtensionsServiceProviderFactory()
				.UseMauiApp<App>()
				.ConfigureFonts(fonts =>
				{
					fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
				})
				.ConfigureServices(services =>
				{
					services.AddBlazorWebView();
					services.AddBlazorStuff("http://10.0.2.2:5000/");
				});
		}
	}
}
```

**Comment by themronion** [comment]
Did u configure the android manifest to allow clear traffic?

**Comment by ElliottBrand** [comment]
I hadn't, but I just tried it and it still crashed
```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android">
	<uses-sdk android:minSdkVersion="21" android:targetSdkVersion="30" />
	<application android:allowBackup="true" android:icon="@mipmap/appicon" android:usesCleartextTraffic="true" android:roundIcon="@mipmap/appicon_round" android:supportsRtl="true"></application>
	<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
</manifest>
```

I also tried adding ```<uses-permission android:name="android.permission.INTERNET" />``` just to see if it made any difference, but it didn't.

**Comment by icebeam7** [comment]
I faced the same situation. I have an asynchronous method (check [here](https://github.com/icebeam7/CountryApp/blob/master/CountryApp/CountryApp/Services/GeonamesService.cs)). 

I solved it by wrapping it inside a Task.Run and getting the .Result object. Check [here](https://github.com/icebeam7/CountryApp/blob/master/CountryApp/CountryApp/ViewModels/CountryListViewModel.cs#L26)

I know it's not the best approach...

**Comment by ElliottBrand** [comment]
Thanks for the response, but this doesn't seem to stop it from crashing for me. I'm guessing the fact that I'm pulling the data into a Blazor component may be the difference.

I have to launch this using ```dotnet build -t:Run -f net6.0-android```. Is there a way to launch a debugger (similar to how 'dotnet watch run' would work in a web app)? Maybe it would produce some details in command line that could help diagnose the problem.

**Comment by ElliottBrand** [comment]
So, after reading https://devblogs.microsoft.com/xamarin/cleartext-http-android-network-security/ it does seem like @themronion may correct and that it must be related to cleartext http, though I haven't been able to get it to work. When I'm trying to use http://10.0.2.2:5000/ as the BaseUri, it's crashing exactly like it does when pointing to a secure URL. It's like 10.0.2.2 is defaulting to https for the API, but not when I navigate to http://10.0.2.2:5000/api/about in the emulator web browser (which displays the Json data on the page). The strange thing is when I pointed my app to an external API (from the Blazor Weather App example) that's at a non-secure http URL, it received the data just fine from that API...without any adjustments to the AndroidManifest.xml file to tell it to allow cleartext http.

**Comment by ElliottBrand** [comment]
I was able to track down my issue here. It was unrelated to MAUI. Even though I was using http://10.0.2.2:5000 for the BaseUri of the HttpClient, which was the insecure path, it was redirecting to https and port 5001 on the server because the Startup project in the server was Configured with ```app.UseHttpsRedirection();```  I just moved that into an if statement, so it only worked if not in the development environment.

```
if (env.IsDevelopment())
{
    ...
}
else
{
    ...
    app.UseHttpsRedirection();
}
```
The HttpClient now receives the data from the API.
